### PR TITLE
Remove redundant AddressConfigType

### PR DESCRIPTION
**Pull request by manadart**
For some time we've had the types `AddressConfigType` and `AddressConfigMethod` kicking around. These are essentially duplicates with the exception of one using "dynamic" (from [man page](https://manpages.debian.org/jessie/ifupdown/interfaces.5.en.html)) where the other uses "dhcp".

Since we store addresses with the latter type, we adopt the strategy of removing the former so as to avoid the need for a DB upgrade step.

A facade bump ensures that we maintain compatibility in the provisioner API for older agents brokering containers.

Most of the patch is name changes. I have sought to comment the pertinent parts.

## QA steps

For these scenarios:
- Bootstrapped with this patch.
- Bootstrapped with 2.8 and a controller (only) upgraded to this patch;

Check that machine and container-in-machine provisioning work on:
  - MAAS (container-networking-method=provider)
  - AWS (container-networking-method=fan)

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

File: apiserver/facades/agent/provisioner/provisioner_test.go [hunk 1]
```diff
@@ -2001,24 +1998,6 @@ func (s *provisionerMockSuite) expectLinkLayerDevices() {
 	pExp.Name().Return(devName).MinTimes(1)
 }
 
-func (s *provisionerMockSuite) TestContainerAlreadyProvisionedError(c *gc.C) {
```
Comment (manadart) [inline apiserver/facades/agent/provisioner/provisioner_test.go:2004]: This code-path has not been used for some time and is now impossible to invoke.

File: apiserver/params/network.go [hunk 2]
```diff
@@ -152,10 +152,9 @@ type NetworkConfig struct {
 	// auto-start.
 	NoAutoStart bool `json:"no-auto-start,omitempty"`
 
-	// ConfigType, if set, defines what type of configuration to use.
-	// See network.AddressConfigType for more info. If not set, for
-	// backwards-compatibility, "dhcp" is assumed.
-	ConfigType string `json:"config-type,omitempty"`
+	// ConfigMethod defines what type of configuration to use.
+	// See network.AddressConfigMethod for more info.
+	ConfigMethod string `json:"config-type,omitempty"`
```
Comment (manadart) [inline apiserver/params/network.go:157]: Note that the name is updated for congruence, but the tag preserves backward-compatibility.

File: apiserver/common/networkingcommon/types.go [hunk 3]
```diff
@@ -239,23 +239,16 @@ func networkAddressToStateArgs(
 		return state.LinkLayerDeviceAddress{}, errors.Trace(err)
 	}
 
-	var derivedConfigMethod network.AddressConfigMethod
-	switch method := network.AddressConfigMethod(dev.ConfigType); method {
-	case network.StaticAddress, network.DynamicAddress,
-		network.LoopbackAddress, network.ManualAddress:
-		derivedConfigMethod = method
-	case "dhcp": // awkward special case
```
Comment (manadart) [inline apiserver/common/networkingcommon/types.go:247]: This conversion is now effectively done in `params.InterfaceInfoFromNetworkConfig`

File: cloudconfig/cloudinit/network_ubuntu.go [hunk 4]
```diff
@@ -260,8 +260,8 @@ func PrepareNetworkConfigFromInterfaces(interfaces corenetwork.InterfaceInfos) (
 		}
 		if cidr != "" {
 			nameToAddress[ifaceName] = cidr
-		} else if info.ConfigType == corenetwork.ConfigDHCP {
-			nameToAddress[ifaceName] = string(corenetwork.ConfigDHCP)
+		} else if info.ConfigMethod == network.DynamicAddress {
```
Comment (manadart) [inline cloudconfig/cloudinit/network_ubuntu.go:263]: Here is the reason we needed a provisioner API bump for backward compatibility.

The only place that backward compatibility was broken was for container interface info sent by the provisioner to local container brokers. Those would be expecting "dhcp" instead or "dynamic".

**Comment by manadart** [comment]
[Remove duplicated AddressConfigType](https://trello.com/c/Tu5p0H7V/3645-remove-duplicated-addressconfigtype)

**Review by achilleasa — APPROVED** [review]
LGTM.

Just some notes:
1) It would be much easier to review this if the corenetwork -> network mechanical changes got tackled in a separate PR.
2) We should probably ping the folks working on vsphere to rebase their changes after this lands.

**Comment by manadart** [comment]
Superseded by #12947

**Pull request closed without merging**